### DISCUSSION TITLE
Display energy output / efficiency / COP

### DISCUSSION THREAD

**Discussion by jh537** [emsesp/EMS-ESP32#2137] [Ideas]
Hey,
you can display the compressor power in real time. Couldn't you also calculate the current energy output and then use this to calculate the COP or efficiency? I would really like that.

**Comment by MichaelDvP** [comment]
> Couldn't you also calculate the current energy output and then use this to calculate the COP or efficiency?

I don't have a heatpump, but as far as i see we don't have the values to calculate this. We have the input power and the in/out temperatures, but no flow to calculate the thermal power. Can you give a hint how to calculate output power.

**Comment by jh537** [comment]
There are a lot of energy consumption and generation values. These can be recognized as counter values in ESP. They are constantly changing. If you take the values in a time period, e.g. one minute, and calculate the difference, you should also be able to calculate the energy output. Otherwise there is also the flow rate of the two pumps PC0 and PC1 per hour to evaluate. I have a CS5800 here Does that help?

**Comment by jimmyhenne** [comment]
Some heatpumps have the values to calculate.
For example: Buderus wlw196i: in EMS-ESP there is _boiler_nrgsupptotal_ and _boiler_nrgconstotal_. There are some more sensors where the values are seperated in heating and warmwater production.
COP calculating would be _boiler_nrgsupptotal_ / _boiler_nrgconstotal_
One has to know that especially the value for the electrical values are not accurate, they are estimated.

When your heating delivers the values, you can calculate COP by yourself in Home Assistant (and probably also in other smarthome systems) with a template helper.

**Comment by jh537** [comment]
![grafik](https://github.com/user-attachments/assets/be8042ae-3c9d-41ad-adec-5e0e457ec703)

The current output of the heat pump is now displayed in EMS. The value for the output power is only available as a cumulative meter value. I have not yet found a way to convert this value into an instantaneous value. I have also looked at Home Assistant.  It will be a while before I can perhaps implement this. Other users may not be able to calculate the value manually. For this reason, it would be very interesting and useful in the ESP.

**Comment by jimmyhenne** [comment]
Do you have these entities?
![grafik](https://github.com/user-attachments/assets/379a1974-5074-4155-bc1b-f87da93966ee)

You don't need an instantaneous value for calculating the COP. Does your screenshot show the output heating power (_boiler_nrgsuppheating_) or the consumpted energy power (_boiler_nrgconscompheating_)?

The total energy output is just fine. You only need a total output heating power in kWh and a total electrical energy consumed in kwH.

**Reply by jh537** [reply]
Of course I have these values, I can also display them in the app - but I'm interested in the current live values

**Comment by jimmyhenne** [comment]
Ok, with these values you can calculate COP in total (lets say lifetime), yearly, monthly, daily or hourly. But I think you can't calculate it live.
I'm not that deep into the definitions but I think COP is defined as heat generated / electricity required and the unit of both is consumption stated in kWh so wihtin in an hour. Live values would be kW (without h) I think.
You cannot calculate from generated heat or consumption (kwh) back to the current power (kw), because the power is not constant.
When you want to know the actual energy consumption you will need a smart energy meter like a SDM630 Modbus that can be integrated to Home Assistant as well. It can also measure the energy comsumption more accurate that ne heat pump. But the actual heat generation would still be missing.

**Reply by jh537** [reply]
with a heat meter it should definitely be accurate and also work...  I still have no idea what can be used for this at a reasonable price

**Reply by jimmyhenne** [reply]
You are right. A smart heat meter could our would provide the value, but I don't know if there are any available which can be integrated in HA.

**Reply by jh537** [reply]
It can certainly also be calculated from the flow and return temperatures and the flow rate. All these values are available. The current consumption is now also displayed. This would then make it possible to optimize efficiency, e.g. by changing the pump pressure.
